docs/lazy-adr/adr-009-state-fraud-proofs.md

```diff
@@ -0,0 +1,234 @@
+  int64 block_height = 1;
+  // Intermediate State Root right before the fraudulent state transition
+  bytes app_hash = 2;
+
+  // Map from an app module name to a State Witness
+  map<string, StateWitness> state_witness = 3;
+
+  // Fraudulent state transition has to be one of these
+  // Only one have of these three can be non-nil
+  RequestBeginBlock fraudulent_begin_block = 4;
+  RequestDeliverTx fraudulent_deliver_tx = 5;
+  RequestEndBlock fraudulent_end_block = 6;
+}
+
+// State witness with a list of all witness data
+message StateWitness {
+  // store level merkle inclusion proof
+  tendermint.crypto.ProofOp proof = 1;
+  // merkle root hash of the substore
+  bytes root_hash = 2;
+  // List of witness data
+  repeated WitnessData witness_data = 3;
+}
+
+// Witness data containing a key/value pair and a Merkle inclusion proof for said key/value pair
+message WitnessData {
+  bytes key = 1;
+  bytes value = 2;
+  // substore level merkle inclusion proof
+  tendermint.crypto.ProofOp proof = 3;
+}
+```
+
+Finally, return this generated state fraud proof back to the Rollmint Full Node:
+
+```protobuf
+message ResponseGenerateFraudProof {
+  FraudProof fraud_proof = 1;
+}
+```
+
+### Gossiping Fraud Proofs
+
+After a Rollmint Full Node generates a Fraud Proof, it gossips the Fraud Proof over P2P to Rollmint light clients. 
+
+### Verifying Fraud Proofs
+
+Rollmint light clients should be able to use these gossiped state fraud proofs to verify whether a fraudulent state transition occured or not by themselves. 
+
+There are four stages of verification that must occur for a Fraud Proof. The first three stages take place in Rollmint and verify that the fraud proof itself was generated correctly. The fourth stage takes place in a Cosmos SDK app to actually execute the fraudulent state transition embedded in the fraud proof.
+
+**Stage One**
+
+Verify that both the `appHash` (ISR) and the fraudulent state transition in the `FraudProof` exist as part of a block published on the DA layer within a specified fraud proof window.
+
+**Stage Two**
+
+Go through the `state_witness` list in the `FraudProof` and verify that all the store level merkle inclusion proofs are valid: the corresponding `root_hash` was included in a merkle tree with root `appHash`.
+
+**Stage Three**
+
+Go through the `WitnessData` in each `StateWitness` and verify that all the substore level merkle inclusion proofs are valid: the corresponding `key` was included in a merkle tree with root `root_hash`.
+
+**Stage Four**
+
+Spin up a new Cosmos SDK app and initialize its store with Deep Subtrees constructed using merkle inclusion proofs in the `FraudProof`. After this initialization, the app hash representing the state of the app should match the `appHash` inside the `FraudProof`. This store should now contain all the key/value pairs and underlying tree structure necessary to execute the fraudulent state transition contained in the `FraudProof`. 
```

Comment:
[non-blocking]
I suggest using a bit more general term in here, as implementation may contain "non-inclusion-proofs", etc.

```suggestion
Spin up a new Cosmos SDK app and initialize its store with Deep Subtrees constructed using witness data in the `FraudProof`. After this initialization, the app hash representing the state of the app should match the `appHash` inside the `FraudProof`. This store should now contain all the key/value pairs and underlying tree structure necessary to execute the fraudulent state transition contained in the `FraudProof`. 
```